public/js/reducers/sources.js

```diff
@@ -54,19 +54,19 @@ function update(state = State(), action: Action) : Record<SourcesState> {
 
     case "SELECT_SOURCE":
       return state.merge({
-        selectedSource: action.source,
         pendingSelectedSourceURL: null,
-        tabs: updateTabList(state, fromJS(action.source), action.options)
-      });
+        tabs: updateTabList(state, fromJS(action.source), action.tabIndex)
+      }).set("selectedLocation", { sourceId: action.source.id,
+                                   line: action.line });
 
     case "SELECT_SOURCE_URL":
       return state.merge({ pendingSelectedSourceURL: action.url });
 
     case "CLOSE_TAB":
-      return state.merge({
-        selectedSource: getNewSelectedSource(state, action.id),
-        tabs: removeSourceFromTabList(state, action.id)
-      });
+      return state.merge({ tabs: removeSourceFromTabList(state, action.id) })
+        .set("selectedLocation", {
+          sourceId: getNewSelectedSource(state, action.id)
+        });
```

Comment:
Unfortunately, `merge` basically calls `toJS` and deeply converts everything to an immutable object. I wanted to avoid more `getIn` calls, because eventually we'll be using Records more thoroughly so we could just use `merge` here but it would still be `selectedLocation.line` elsewhere. So this is temporary until we figure out how to use Records better.

But now that it's implemented, I can look over and see if it's simpler to keep it as immutable.